AppCenter/AppCenter/Internals/Storage/MSDBStorage.m

```diff
@@ -7,67 +7,116 @@
-
-  // TODO: Introduce a versioning / data migration system in case of updates on the schema.
+- (instancetype)initWithSchema:(MSDBSchema *)schema version:(NSUInteger)version filename:(NSString *)filename {
   if ((self = [super init])) {
-    NSMutableArray *tableQueries = [NSMutableArray new];
-    NSMutableDictionary *dbColumnsIndexes = [NSMutableDictionary new];
 
     // Path to the database.
     _dbFileURL = [MSUtility createFileAtPathComponent:filename withData:nil atomically:NO forceOverwrite:NO];
 
     // If it is custom SQLite library we need to turn on URI filename capability.
     sqlite3_config(SQLITE_CONFIG_URI, 1);
 
-    // Browse tables.
-    for (NSString *tableName in schema) {
-      NSMutableDictionary *tableColumnsIndexes = [NSMutableDictionary new];
-      NSMutableArray *columnQueries = [NSMutableArray new];
-      NSArray<NSDictionary *> *columns = schema[tableName];
-
-      // Optimization, don't even compute the query if the table already exists.
-      BOOL tableDontExist = ![self tableExists:tableName];
+    // Execute all initialize operation with one database instance.
+    [self executeWithDatabase:^int(void *db) {
+      
+      // Create tables based on schema.
+      NSUInteger tablesCreated = [MSDBStorage createTablesWithSchema:schema inDatabase:db];
+      
+      BOOL newDatabase = tablesCreated == schema.count;
+      NSUInteger databaseVersion = [MSDBStorage versionInDatabase:db];
+      if (databaseVersion < version && !newDatabase) {
+        MSLogInfo([MSAppCenter logTag], @"Migrate \"%@\" database from %lu to %lu version.",
+                   filename, (unsigned long)databaseVersion, (unsigned long)version);
+        [self migrateDatabase:db fromVersion:databaseVersion];
+      }
+      [MSDBStorage setVersion:version inDatabase:db];
+      
+      return SQLITE_OK;
+    }];
+  }
+  return self;
+}
 
-      // Browse columns.
-      for (NSUInteger i = 0; i < columns.count; i++) {
-        NSString *columnName = columns[i].allKeys[0];
-        [tableColumnsIndexes setObject:@(i) forKey:columnName];
+- (BOOL)executeWithDatabase:(int (^)(void *))callback {
+  sqlite3 *db = NULL;
+  int result = SQLITE_OK;
+  result = sqlite3_open_v2([[self.dbFileURL absoluteString] UTF8String], &db,
+                           SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_URI, NULL);
+  if (result == SQLITE_OK) {
+    result = callback(db);
+  } else {
+    MSLogError([MSAppCenter logTag], @"Failed to open database with result: %d.", result);
+  }
+  sqlite3_close(db);
+  return SQLITE_OK == result;
+}
 
-        // Compute column query.
-        if (tableDontExist) {
-          [columnQueries addObject:[NSString stringWithFormat:@"\"%@\" %@", columnName,
-                                                              [columns[i][columnName] componentsJoinedByString:@" "]]];
-        }
-      }
++ (NSUInteger)createTablesWithSchema:(MSDBSchema *)schema inDatabase:(void *)db {
```

Comment:
Please rename `inDatabase` by `inOpenedDatabase`, we want to make it explicit that the database has to be opened before it's passed to this method.